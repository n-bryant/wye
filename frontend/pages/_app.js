import React from "react";
import App from "next/app";

import Page from "../src/components/Page";
import { ApolloProvider } from "react-apollo";
import withData from "../lib/withData";

import CssBaseline from "@material-ui/core/CssBaseline";

import NextNProgress from "../src/components/NextNProgress";

// set up a context provider
const AppContext = React.createContext({});
export const AppContextProvider = AppContext.Provider;
export const AppContextConsumer = AppContext.Consumer;

// reducer action and content options
export const ACTIONS = {
  SET_USERS: "setUsers",
  SET_CONTENT: "setContent"
};
export const CONTENT_OPTIONS = {
  WELCOME: "WELCOME",
  FORM: "FORM",
  RECOMMENDATIONS: "RECOMMENDATIONS"
};

// initial state values
export const INITIAL_STATE = {
  content: CONTENT_OPTIONS.WELCOME,
  users: []
};

/**
 * reducer for updating state values
 * @param {Object} state
 * @param {Object} action
 * @returns {Object}
 */
export const reducer = (state, action) => {
  switch (action.type) {
    case ACTIONS.SET_USERS:
      return {
        ...state,
        users: action.value
      };
    case ACTIONS.SET_CONTENT:
      return {
        ...state,
        content: action.value
      };
    default:
      throw new Error();
  }
};

export class WyeApp extends App {
  componentDidMount() {
    // Remove the server-side injected CSS.
    const jssStyles = document.querySelector("#jss-server-side");
    if (jssStyles) {
      jssStyles.parentElement.removeChild(jssStyles);
    }
  }

  static async getInitialProps({ Component, ctx }) {
    let pageProps = {};
    if (Component.getInitialProps) {
      pageProps = await Component.getInitialProps(ctx);
    }
    // expose query to the user
    pageProps.query = ctx.query;
    return { pageProps };
  }

  render() {
    return <WyeWithState {...this.props} />;
  }
}

// app with state
export const WyeWithState = props => {
  const { Component, apollo, pageProps } = props;
  const [state, dispatch] = React.useReducer(reducer, INITIAL_STATE);

  return (
    <React.Fragment>
      <CssBaseline />
      <NextNProgress />
      <ApolloProvider client={apollo}>
        <AppContextProvider
          value={{
            state,
            dispatch
          }}
        >
          <Page>
            <Component {...pageProps} />
          </Page>
        </AppContextProvider>
      </ApolloProvider>
    </React.Fragment>
  );
};

// app with apollo client
export const WyeWithData = withData(WyeApp);

export default WyeWithData;
