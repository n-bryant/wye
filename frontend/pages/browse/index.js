import React from "react";
import PropTypes from "prop-types";

import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../lib/classNamePrefix";
import { withWidth } from "@material-ui/core/";
import { withStyles } from "@material-ui/core/styles";

import Container from "@material-ui/core/Container";
import Typography from "@material-ui/core/Typography";
import BackgroundProvider from "../../src/components/BackgroundProvider";
import styles from "./index.styles";

/**
 * renders the Browse page
 */
export const BrowsePage = props => {
  const classnames = BrowsePage.classnames(props);

  //                              top right of page ->     advanced search
  // Heading - Browse Multiplayer Games
  // Subheading - Browse the most popular and discounted multiplayer games

  // trending - most popular by recent playtime
  // - Cards: Contain appropriate capsule image, name, and price details
  //    - on hover: shows name, user rating, dev/pub info and
  // capsuleLg - 1:4

  // publishers - games by most popular publishers
  // - Cards: 2 pages of 4 cards; page paddles and page pills
  //  - Header image linking to most popular title by publisher
  //  - Publisher name linking to /publisher/[name]

  // big sales - top 4 - rated by discount and user rating
  // - Cards: 2 pages of 3 cards; page paddles and page pills
  //  - libraryCapsule image on top
  //  - name / sale details on bottom
  //  - on hover: - name, user rating, user details if available (wanted by)

  // buttons to go to filter for under $5 / $10

  // Browse by Genre
  // - grid of genre buttons

  return (
    <BackgroundProvider>
      <div className={classnames.root()}>
        <Container
          className={classnames.element("container")}
          maxWidth="lg"
          disableGutters
        >
          <Container className={classnames.element("main")} maxWidth={false}>
            <Typography className={classnames.element("heading")} variant="h1">
              Featured Games
            </Typography>
            <Typography
              className={classnames.element("subHeading")}
              variant="h2"
            >
              Browse the most popular and discounted multiplayer games
            </Typography>
          </Container>
        </Container>
      </div>
    </BackgroundProvider>
  );
};
BrowsePage.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}BrowsePage`
);
BrowsePage.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    container: PropTypes.string,
    main: PropTypes.string,
    heading: PropTypes.string,
    subHeading: PropTypes.string
  })
};
BrowsePage.defaultProps = {
  classes: {}
};

export default withWidth()(withStyles(styles)(BrowsePage));
