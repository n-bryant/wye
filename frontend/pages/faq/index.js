import React from "react";
import PropTypes from "prop-types";

import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../lib/classNamePrefix";
import { withStyles } from "@material-ui/core/styles";

import FaqAccordion from "../../src/components/FaqAccordion";
import styles from "./index.styles";

/**
 * renders an FAQ page
 */
export const FaqPage = props => {
  const classnames = FaqPage.classnames(props);

  return (
    <div className={classnames.root()}>
      <FaqAccordion />
    </div>
  );
};
FaqPage.classnames = createClassNameHelper(`${JSS_CLASS_NAME_PREFIX}FaqPage`);
FaqPage.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string
  })
};
FaqPage.defaultProps = {
  classes: {}
};

export default withStyles(styles)(FaqPage);
