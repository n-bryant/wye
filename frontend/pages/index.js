import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../lib/classNamePrefix";

import Icon from "@mdi/react";
import { mdiGithubCircle } from "@mdi/js";
import { AppBar, Container, Toolbar, Typography } from "@material-ui/core";

import Link from "next/link";
import WelcomeEmptyState from "../src/components/WelcomeEmptyState";
import RecommendationsForm from "../src/components/RecommendationsForm/RecommendationsForm";
import IndexPageBackground from "../src/components/ScrollingBackground/IndexPageBackground";
import styles from "./index.styles";

// set up a context provider
const AppContext = React.createContext({});
export const AppContextProvider = AppContext.Provider;
export const AppContextConsumer = AppContext.Consumer;

// initial state values
export const INITIAL_STATE = {
  showForm: false,
  users: []
};
/**
 * reducer for updating state values
 * @param {Object} state
 * @param {Object} action
 * @returns {Object}
 */
export const reducer = (state, action) => {
  switch (action.type) {
    case "setShowForm":
      return {
        ...state,
        showForm: action.value
      };
    case "setUsers":
      return {
        ...state,
        users: action.value
      };
    default:
      throw new Error();
  }
};

/**
 * renders the index page's content,
 * dynamically displaying either the welcome content or the recommendations form
 */
export const Index = props => {
  const classnames = Index.classnames(props);
  const [state, dispatch] = React.useReducer(reducer, INITIAL_STATE);

  return (
    <AppContextProvider
      value={{
        state,
        dispatch
      }}
    >
      <div className={classnames.root()}>
        <Container
          className={classnames.element("container")}
          maxWidth={false}
          disableGutters
        >
          <AppBar position="static" className={classnames.element("appBar")}>
            <Toolbar className={classnames.element("toolBar")}>
              <Typography className={classnames.element("title")} variant="h1">
                Wye
              </Typography>
              <div className={classnames.element("linksContainer")}>
                <Link href="/faq">
                  <a className={classnames.element("link")}>faq</a>
                </Link>
                <Link href="https://github.com/n-bryant/wye" prefetch={false}>
                  <a className={classnames.element("link")} target="_blank">
                    <Icon
                      className={classnames.element("githubIcon")}
                      path={mdiGithubCircle}
                    />
                  </a>
                </Link>
              </div>
            </Toolbar>
          </AppBar>
          <div className={classnames.element("main")}>
            <div className={classnames.element("contentContainer")}>
              <div className={classnames.element("content")}>
                {state.showForm ? (
                  <RecommendationsForm className={classnames.element("form")} />
                ) : (
                  <WelcomeEmptyState />
                )}
              </div>
            </div>
          </div>
        </Container>
        <IndexPageBackground />
      </div>
    </AppContextProvider>
  );
};
Index.classnames = createClassNameHelper(`${JSS_CLASS_NAME_PREFIX}Index`);
Index.propTypes = {
  // styles applied
  classes: PropTypes.shape({
    root: PropTypes.string,
    container: PropTypes.string,
    appBar: PropTypes.string,
    toolBar: PropTypes.string,
    title: PropTypes.string,
    linksContainer: PropTypes.string,
    link: PropTypes.string,
    githubIcon: PropTypes.string,
    form: PropTypes.string,
    main: PropTypes.string,
    contentContainer: PropTypes.string,
    content: PropTypes.string
  })
};
Index.defaultProps = {
  classes: {}
};

export default withStyles(styles)(Index);
