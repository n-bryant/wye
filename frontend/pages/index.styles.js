const styles = theme => ({
  root: {
    height: "100%",
    position: "relative",
    overflow: "hidden"
  },
  container: {
    backgroundColor: "rgba(34, 139, 230,.25)",
    height: "100%",
    zIndex: "1"
  },
  main: {
    height: "100%",
    overflow: "auto"
  },
  contentContainer: {
    height: "100%",
    width: "100%",
    maxWidth: "1500px",
    minHeight: "530px",
    margin: "0 auto",
    padding: "3rem",
    [theme.breakpoints.down("xs")]: {
      padding: "0"
    }
  },
  content: {
    height: "100%",
    width: "100%",
    background: "rgba(0,0,0,0.8)",
    backdropFilter: "blur(6px)",
    border: "1px solid hsla(0,0%,100%,0.2)",
    overflow: "auto"
  }
});
export default styles;
