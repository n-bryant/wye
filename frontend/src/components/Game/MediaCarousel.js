import React from "react";
import PropTypes from "prop-types";

import { withStyles } from "@material-ui/core/styles";
import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";

import Grid from "@material-ui/core/Grid";
import Button from "@material-ui/core/Button";

import Icon from "@mdi/react";
import { mdiPlay } from "@mdi/js";

import styles from "./MediaCarousel.styles";

export const MEDIA_TYPES = {
  VIDEO: "video",
  IMG: "img"
};

/**
 * renders a media carousel with a featured item
 */
export const MediaCarousel = props => {
  const classnames = MediaCarousel.classnames(props);
  const { media } = props;

  const initialState = {
    item: media[0],
    type: media[0].title ? MEDIA_TYPES.VIDEO : MEDIA_TYPES.IMG
  };
  const [featuredItem, setFeaturedItem] = React.useState(initialState);

  // build an img or video based on the featuredItem's type
  const Feature = () => {
    const { item } = featuredItem;
    if (featuredItem.type === MEDIA_TYPES.VIDEO) {
      return (
        <video
          className={classnames.element("featuredItem")}
          playsInline={true}
          src={item.fullsizeUrl}
          poster={item.thumbnail}
          autoPlay
          preload={"none"}
          controls
          muted
        >
          {item.title}
        </video>
      );
    } else {
      const [loaded, setLoaded] = React.useState(false);

      const onLoad = () => {
        setLoaded(true);
      };
      return (
        <img
          className={classnames.element("featuredItem", {
            hidden: !loaded
          })}
          src={item.fullsizeUrl}
          alt="featured item"
          onLoad={onLoad}
        />
      );
    }
  };

  return (
    <Grid
      className={classnames.root()}
      container
      item
      xs={12}
      md={7}
      spacing={1}
    >
      <Grid
        className={classnames.element("featuredItemContainer")}
        item
        xs={12}
      >
        <div className={classnames.element("featuredItemWrapper")}>
          <img
            className={classnames.element("featuredItemSpacer")}
            src="/featuredItemSpacer.png"
          />
          <Feature />
        </div>
      </Grid>
      <Grid
        className={classnames.element("thumbnailsContainer")}
        container
        wrap="nowrap"
        item
        xs={12}
        spacing={1}
      >
        {media.map(item => (
          <Grid
            key={item.id}
            item
            className={classnames.element("setFeaturedButtonContainer")}
          >
            <Button
              className={classnames.element("setFeaturedButton")}
              onClick={() =>
                setFeaturedItem({
                  item,
                  type: item.title ? MEDIA_TYPES.VIDEO : MEDIA_TYPES.IMG
                })
              }
            >
              <img
                className={classnames.element("thumbnail")}
                src={item.thumbnailUrl}
                alt={item.id}
              />
              {item.title && (
                <div className={classnames.element("playIconContainer")}>
                  <Icon
                    className={classnames.element("playIcon")}
                    path={mdiPlay}
                  />
                </div>
              )}
            </Button>
          </Grid>
        ))}
      </Grid>
    </Grid>
  );
};
MediaCarousel.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}MediaCarousel`
);
MediaCarousel.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    featuredItemContainer: PropTypes.string,
    featuredItemSpacer: PropTypes.string,
    featuredItem: PropTypes.string,
    featuredItemHidden: PropTypes.string,
    featuredItemWrapper: PropTypes.string,
    thumbnailsContainer: PropTypes.string,
    setFeaturedButtonContainer: PropTypes.string,
    setFeaturedButton: PropTypes.string,
    thumbnail: PropTypes.string,
    playIconContainer: PropTypes.string,
    playIcon: PropTypes.string
  }),
  // details for the media to display
  media: PropTypes.array
};
MediaCarousel.defaultProps = {
  classes: {}
};

export default withStyles(styles)(MediaCarousel);
