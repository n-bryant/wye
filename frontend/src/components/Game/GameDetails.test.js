import GameDetailsWithContext, { GameDetails } from "./GameDetails";
import MediaCarousel from "./MediaCarousel";
import { ACTIONS, CONTENT_OPTIONS } from "../../../pages/_app";

jest.mock("next/router", () => ({
  useRouter: () => ({
    push: jest.fn()
  })
}));

describe("GameDetailsWithContext", () => {
  it("should pass a users value and a dispatch value from context to its rendered child", () => {
    const renderCallbackArgs = {
      state: {
        users: ["foo"]
      },
      dispatch: jest.fn()
    };
    const props = {
      fizz: "buzz"
    };

    const wrapper = shallow(<GameDetailsWithContext {...props} />);
    expect(wrapper.prop("children")(renderCallbackArgs).props.users).toBe(
      renderCallbackArgs.state.users
    );
    expect(wrapper.prop("children")(renderCallbackArgs).props.dispatch).toBe(
      renderCallbackArgs.dispatch
    );
    expect(wrapper.prop("children")(renderCallbackArgs).props.fizz).toBe(
      props.fizz
    );
  });
});

describe("GameDetails", () => {
  const props = {
    classes: {
      root: "root",
      loaded: "loaded",
      backgroundPlaceholder: "backgroundPlaceholder",
      detailsContainer: "detailsContainer",
      detailsContainerWithMaxWidth: "detailsContainerWithMaxWidth",
      recommendationsButton: "recommendationsButton",
      icon: "icon"
    },
    data: {
      details: {
        highlightedVideos: [{}],
        screenshots: [],
        videos: []
      },
      articles: []
    },
    users: [],
    dispatch: jest.fn(),
    width: "md"
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render successfully", () => {
    const wrapper = shallow(<GameDetails {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should set the content to the form when the recommendations button is clicked if there are no users", () => {
    const wrapper = shallow(<GameDetails {...props} />);
    wrapper
      .findWhere(n =>
        n.hasClass(`wye-GameDetails__${props.classes.recommendationsButton}`)
      )
      .prop("onClick")();
    expect(props.dispatch).toHaveBeenCalledWith({
      type: ACTIONS.SET_CONTENT,
      value: CONTENT_OPTIONS.FORM
    });
  });

  it("should set the content to recommendations when the recommendations button is clicked if there are users", () => {
    const wrapper = shallow(<GameDetails {...props} />);
    wrapper.setProps({
      users: ["foo"]
    });
    wrapper.update();
    wrapper
      .findWhere(n =>
        n.hasClass(`wye-GameDetails__${props.classes.recommendationsButton}`)
      )
      .prop("onClick")();
    expect(props.dispatch).toHaveBeenCalledWith({
      type: ACTIONS.SET_CONTENT,
      value: CONTENT_OPTIONS.RECOMMENDATIONS
    });
  });

  it("should not render the MediaCarousel if there is no media to render", () => {
    const wrapper = shallow(<GameDetails {...props} />);
    wrapper.setProps({
      data: {
        details: {
          ...props.data.details,
          highlightedVideos: []
        }
      }
    });
    wrapper.update();
    expect(wrapper.find(MediaCarousel).length).toBe(0);
  });
});
