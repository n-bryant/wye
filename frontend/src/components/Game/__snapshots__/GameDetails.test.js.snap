// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameDetails should render successfully 1`] = `
<BackgroundQuery>
  <div
    className="wye-GameDetails makeStyles-root-1"
  >
    <WithStyles(ForwardRef(Grid))
      className="wye-GameDetails__detailsContainer wye-GameDetails__detailsContainer--withMdWidth makeStyles-detailsContainer-2 makeStyles-detailsContainerWithMdWidth-3"
      container={true}
      spacing={1}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        lg={8}
        xs={12}
      >
        <WithStyles(ForwardRef(Button))
          className="wye-GameDetails__recommendationsButton makeStyles-recommendationsButton-9"
          onClick={[Function]}
        >
          <Icon
            className="wye-GameDetails__icon makeStyles-icon-10"
            color={null}
            horizontal={false}
            path="M20,9V15H12V19.84L4.16,12L12,4.16V9H20Z"
            rotate={0}
            size={null}
            spin={false}
            vertical={false}
          />
          <WithStyles(ForwardRef(Typography))
            variant="body2"
          >
            recommendations
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={12}
      >
        <WithStyles(ForwardRef(Typography))
          variant="h1"
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="wye-GameDetails__detailsInnerContainer makeStyles-detailsInnerContainer-4"
        container={true}
        item={true}
        justify="space-evenly"
        spacing={1}
        xs={12}
      >
        <WithStyles(MediaCarousel)
          classes={Object {}}
          media={
            Array [
              Object {},
            ]
          }
        />
        <WithWidth(WithStyles(ReleaseInfoBlock))
          data={
            Object {
              "highlightedVideos": Array [
                Object {},
              ],
              "screenshots": Array [],
              "videos": Array [],
            }
          }
          mdWidth={5}
        />
      </WithStyles(ForwardRef(Grid))>
      <div
        className="wye-GameDetails__priceContainer makeStyles-priceContainer-7"
      >
        <WithWidth(WithStyles(PriceInfoBlock))
          data={
            Object {
              "metacritic": undefined,
              "price": undefined,
            }
          }
        />
      </div>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(GameArticles)
      articles={Array []}
      classes={Object {}}
    />
  </div>
</BackgroundQuery>
`;
